[1.10] Add log-level option to conmon and crio.conf

umohnani8 opened the pull request:
conmon now has a --log-level flag that changes
the verbosity of the logs based on the value it
is set to. From logrus, panic(0), fatal(1), error(2) tells
conmon to print all pexit and nexit logs (this
is the default). warn(3) adds the nwarn logs, info(4)
adds the ninfo logs, and finally debug(5) adds the ndebug
logs as well.

Adds a log_level field in the crio.conf that can have the
following options: panic, fatal, error, warn, info, and debug.
Changes default from info to error.

Signed-off-by: umohnani8 <umohnani@redhat.com>

umohnani8 commented:
@mrunalp @runcom PTAL

mrunalp commented:
/test all

mheon commented:
Can we get a flag to optionally keep this? It seems valuable for debugging Podman.

umohnani8 commented:
@mrunalp should I go ahead and add a flag? If so, what do we want to call the flag.

rhatdan commented:
Can you just make it debug versus info?

mrunalp commented:
@rhatdan The debug change doesn't work as it just changes what gets printed in the logs. We would need to add a flag to control it.

umohnani8 commented:
Added debug flag. @mrunalp @mheon @rhatdan PTAL

mheon commented on conmon/conmon.c:
```diff
@@ -80,6 +80,18 @@
 	} while (0)
 
+#define ndebug(s)                                             \
+	do {                                                 \
```
Might be better to include the `if(opt_debug)` in the macros so all debug prints will only happen if debug is set.

runcom commented:
Why do you need this for debugging @mheon? Do you have or had any use case that helps you debug with this information? If this is successful the pid is sent back to golang where if in not wrong is printed somewhere, if not, we can add it there.

mheon commented:
Our most successful debug strategy for conmon problems now is to have
people copy the command line libpod would use to launch it and then run it
themselves and see what happens. It lets us get output more reliably than
forwarding to Go with the added benefit that we can easily stick strace on
there if needed to get a better picture of where it's failing. Having the
container pid actually printed is useful there.

On Thu, Jul 12, 2018, 02:43 Antonio Murdaca <notifications@github.com>
wrote:

> Why do you need this for debugging @mheon <https://github.com/mheon>? Do
> you have or had any use case that helps you debug with this information? If
> this is successful the pid is sent back to golang where if in not wrong is
> printed somewhere, if not, we can add it there.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-incubator/cri-o/pull/1692#issuecomment-404406620>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHYHCLt_N4yQmZRq3n7VmzOWFsM0Z-N-ks5uFvAEgaJpZM4VLgF0>
> .
>

runcom commented:
> Our most successful debug strategy for conmon problems now is to have
> people copy the command line libpod would use to launch it and then run it
> themselves and see what happens. It lets us get output more reliably than
> forwarding to Go with the added benefit that we can easily stick strace on
> there if needed to get a better picture of where it's failing.

yeah that's exactly what we do as well to debug conmon, I still can't see how the pid printed after conmon starts is useful, but ok.

umohnani8 commented:
/test all

mrunalp commented:
We decided to add support for log-level instead.

umohnani8 commented:
Added `--log-level` to conmon and changed when the logs are printed based on the level. Also added `log_level` to the crio.conf file.
@mrunalp @runcom PTAL.
Once this gets approved, I will pull in the same thing to master and 1.11.

umohnani8 commented:
/test all

umohnani8 commented:
@TomSweeneyRedHat thanks! Added doc changes.

umohnani8 commented:
/test all

TomSweeneyRedHat commented on tutorial.md:
```diff
@@ -163,7 +163,7 @@ Description=OCI-based implementation of Kubernetes Container Runtime Interface
 
 [Service]
-ExecStart=/usr/local/bin/crio --log-level debug
+ExecStart=/usr/local/bin/crio
```
It would be nice to show an example in crio.conf how to set up the debug level.  Also based on the registry.conf discussions this morning, does the registries option at line 150 still hold?

umohnani8 replied:
That is fine, for 1.10 and below we need to add the default registries in `/etc/crio/crio.conf`.
Added an "optional" to show how we can change the log_level in crio.conf.

rhatdan commented:
Kata does not work on 1.10, so we can ignore that I believe.

mrunalp commented on conmon/conmon.c:
```diff
@@ -171,6 +196,7 @@ static GOptionEntry opt_entries[] =
   { "log-size-max", 0, 0, G_OPTION_ARG_INT64, &opt_log_size_max, "Maximum size of log file", NULL },
   { "socket-dir-path", 0, 0, G_OPTION_ARG_STRING, &opt_socket_path, "Location of container attach sockets", NULL },
   { "version", 0, 0, G_OPTION_ARG_NONE, &opt_version, "Print the version and exit", NULL },
+  { "log-level", 0, 0, G_OPTION_ARG_INT, &opt_log_level, "Print debug logs based on log level", NULL },
```
We should read this as ARG_STRING and map it to int for log_level.

umohnani8 replied:
But then logrus already maps the levels to int, so isn't it better to use that instead of creating a new string to int map of the levels in conmon?

umohnani8 replied:
Fixed. Reading as ARG_STRING and mapping to an int. Not the most high tech way, but it is the simplest solution I could think of. @mrunalp @runcom PTAL.

mrunalp commented on conmon/conmon.c:
```diff
@@ -56,29 +56,53 @@
+	}
+
+#define ndebug(s)                                                     \
+	if (parse_level(opt_log_level) > 2) {                                      \
```
2 doesn't match the enum values you have defined below. We should use the enum constants instead of number in this comparison.

umohnani8 replied:
Fixed.

umohnani8 commented:
/test all

umohnani8 commented:
/test e2e_rhel

umohnani8 commented:
/test e2e_rhel

umohnani8 commented:
/test all

umohnani8 commented:
tests finally pass. @mrunalp @runcom PTAL

runcom commented on cmd/crio/main.go:
```diff
@@ -229,9 +232,9 @@ func main() {
 		cli.StringFlag{
 			Name:  "log-level",
-			Usage: "log messages above specified level: debug, info (default), warn, error, fatal or panic",
+			Value: "error",
```
are we sure about this default? what was before this change overall?

umohnani8 replied:
it was info. We want the logs to be as minimal as possible by default so error seems like the best choice. When we change this in crio.conf the logs get more verbose for both cri-o and conmon.

runcom replied:
awesome, we're good then (error's super fine!)

rhatdan commented:
LGTM
